File: src-tauri/src/models/dto.rs
use serde::{Deserialize, Serialize};

// ---------------------------------------------------------------------------
// DTO types (sent to FE — no internal DB columns exposed)
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FeedDto {
    pub id: i64,
    pub name: String,
    pub url: String,
    pub feed_type: String,
    pub category: String,
    pub enabled: bool,
    pub fetch_interval_minutes: i64,
    pub last_fetched_at: Option<String>,
    pub consecutive_errors: i64,
    pub disabled_reason: Option<String>,
    pub last_error: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, sqlx::FromRow)]
#[serde(rename_all = "camelCase")]
pub struct ArticleDto {
    pub id: i64,
    pub feed_id: i64,
    pub title: String,
    pub url: Option<String>,
    pub summary: Option<String>,
    pub author: Option<String>,
    pub published_at: Option<String>,
    pub importance_score: f64,
    pub is_read: bool,
    pub is_bookmarked: bool,
    pub language: Option<String>,
    pub thumbnail_url: Option<String>,
    pub feed_name: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ArticleDetailDto {
    pub id: i64,
    pub title: String,
    pub url: Option<String>,
    pub content: Option<String>,
    pub summary: Option<String>,
    pub author: Option<String>,
    pub published_at: Option<String>,
    pub feed_name: Option<String>,
    pub importance_score: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DigestDto {
    pub id: i64,
    pub category: String,
    pub title: String,
    pub content_markdown: String,
    pub content_html: Option<String>,
    pub article_count: usize,
    pub model_used: Option<String>,
    pub generated_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UserProfileDto {
    pub display_name: String,
    pub favorite_titles: Vec<String>,
    pub favorite_genres: Vec<String>,
    pub favorite_creators: Vec<String>,
    pub total_read: i64,
}

/// Discover フィード用の拡張 ArticleDto（AI サマリー + スコア付き）
#[derive(Debug, Clone, Serialize, Deserialize, sqlx::FromRow)]
#[serde(rename_all = "camelCase")]
pub struct DiscoverArticleDto {
    pub id: i64,
    pub feed_id: i64,
    pub title: String,
    pub url: Option<String>,
    pub summary: Option<String>,
    pub author: Option<String>,
    pub published_at: Option<String>,
    pub is_read: bool,
    pub is_bookmarked: bool,
    pub language: Option<String>,
    pub thumbnail_url: Option<String>,
    pub feed_name: Option<String>,
    pub ai_summary: Option<String>,
    pub total_score: Option<f64>,
    pub category: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DiscoverFeedResult {
    pub articles: Vec<DiscoverArticleDto>,
    pub total: i64,
    pub has_more: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DeepDiveResult {
    pub question: String,
    pub answer: String,
    pub follow_up_questions: Vec<String>,
    pub provider: String,
    pub citations: Vec<crate::infra::llm_client::Citation>,
}

File: server/src/types/dto.ts
import type { FeedRow } from './models.ts';

// src/types/index.ts に対応する camelCase DTO と、DB 行 (snake_case) → DTO マッパー。

export interface FeedDto {
  id: number;
  name: string;
  url: string;
  feedType: string;
  category: string;
  enabled: boolean;
  fetchIntervalMinutes: number;
  lastFetchedAt: string | null;
  consecutiveErrors: number;
  disabledReason: string | null;
  lastError: string | null;
}

export interface DiscoverArticleDto {
  id: number;
  feedId: number;
  title: string;
  url: string | null;
  summary: string | null;
  author: string | null;
  publishedAt: string | null;
  isRead: boolean;
  isBookmarked: boolean;
  language: string | null;
  thumbnailUrl: string | null;
  feedName: string | null;
  aiSummary: string | null;
  totalScore: number | null;
  category: string | null;
  impactLevel: string | null;
}

export interface DiscoverFeedResult {
  articles: DiscoverArticleDto[];
  total: number;
  hasMore: boolean;
}

export interface ArticleDto {
  id: number;
  feedId: number;
  title: string;
  url: string | null;
  summary: string | null;
  author: string | null;
  publishedAt: string | null;
  importanceScore: number;
  isRead: boolean;
  isBookmarked: boolean;
  language: string | null;
  thumbnailUrl: string | null;
  feedName: string | null;
}

export interface ArticleDetailDto {
  id: number;
  title: string;
  url: string | null;
  content: string | null;
  summary: string | null;
  author: string | null;
  publishedAt: string | null;
  feedName: string | null;
  importanceScore: number;
}

export interface DigestDto {
  id: number;
  category: string;
  title: string;
  contentMarkdown: string;
  contentHtml: string | null;
  articleCount: number;
  modelUsed: string | null;
  generatedAt: string;
}

export interface HighlightEntry {
  article: DiscoverArticleDto;
  reason: string;
}

export interface UnreadCounts {
  forYou: number;
  trending: number;
  anime: number;
  game: number;
  manga: number;
  hardware: number;
  tech: number;
}

export interface KeywordFilterDto {
  id: number;
  keyword: string;
  filterType: string;
  category: string | null;
  createdAt: string;
}

export interface UserProfileDto {
  displayName: string;
  favoriteTitles: string[];
  favoriteGenres: string[];
  favoriteCreators: string[];
  totalRead: number;
}

// --- DB 行型 (snake_case) ---

export interface DiscoverRow {
  id: number;
  feed_id: number;
  title: string;
  url: string | null;
  summary: string | null;
  author: string | null;
  published_at: string | null;
  is_read: number;
  is_bookmarked: number;
  language: string | null;
  thumbnail_url: string | null;
  ai_summary: string | null;
  feed_name: string | null;
  category: string | null;
  impact_level: string | null;
  total_score: number | null;
}

export function toDiscoverArticleDto(r: DiscoverRow): DiscoverArticleDto {
  return {
    id: r.id,
    feedId: r.feed_id,
    title: r.title,
    url: r.url,
    summary: r.summary,
    author: r.author,
    publishedAt: r.published_at,
    isRead: r.is_read !== 0,
    isBookmarked: r.is_bookmarked !== 0,
    language: r.language,
    thumbnailUrl: r.thumbnail_url,
    feedName: r.feed_name,
    aiSummary: r.ai_summary,
    totalScore: r.total_score,
    category: r.category,
    impactLevel: r.impact_level,
  };
}

export function toFeedDto(f: FeedRow): FeedDto {
  return {
    id: f.id,
    name: f.name,
    url: f.url,
    feedType: f.feed_type,
    category: f.category,
    enabled: f.enabled !== 0,
    fetchIntervalMinutes: f.fetch_interval_minutes,
    lastFetchedAt: f.last_fetched_at,
    consecutiveErrors: f.consecutive_errors,
    disabledReason: f.disabled_reason,
    lastError: f.last_error,
  };
}
